#### Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio test. The function should return the F-value and the corresponding p-value for the test.

To calculate the F-value and the corresponding p-value for a variance ratio test, you can use the scipy.stats library, which provides functions for statistical tests. In this case, the `f_oneway` function is suitable for one-way ANOVA, which is used to compare variances across multiple groups.

Here's a Python function that takes two arrays of data and performs the variance ratio test:

In [1]:
import numpy as np
from scipy import stats

def variance_ratio_test(data1, data2):
    """
    Performs a variance ratio test (F-test) to compare the variances of two datasets.
    
    Parameters:
    - data1: Array-like, first dataset
    - data2: Array-like, second dataset
    
    Returns:
    - f_value: The F-statistic value
    - p_value: The p-value associated with the F-statistic
    """
    # Calculate the variances of the two datasets
    var1 = np.var(data1, ddof=1)
    var2 = np.var(data2, ddof=1)
    
    # Calculate the F-value
    f_value = var1 / var2 if var1 > var2 else var2 / var1
    
    # Perform the F-test to get the p-value
    # Degrees of freedom
    df1 = len(data1) - 1
    df2 = len(data2) - 1
    
    # p-value calculation using the F-distribution
    p_value = 1 - stats.f.cdf(f_value, df1, df2)
    
    return f_value, p_value

# Example usage:
data1 = [10, 12, 9, 11, 13]
data2 = [20, 22, 18, 21, 23]

f_value, p_value = variance_ratio_test(data1, data2)
print(f"F-value: {f_value}")
print(f"P-value: {p_value}")


F-value: 1.48
P-value: 0.35665133765231105


Explanation:
- **Calculate Variances**: Compute the variances of the two datasets.
- **F-value**: The F-value is the ratio of the two variances. Ensure the larger variance is in the numerator to get a positive F-value.
- **Degrees of Freedom**: Calculate the degrees of freedom for both datasets.
- **P-value**: Use the F-distribution to compute the p-value, which indicates the probability of observing an F-value at least as extreme as the one computed.

Make sure to replace `data1` and `data2` with your actual datasets when using this function.

#### Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

To compute the critical F-value for a two-tailed test at a given significance level, you need to find the F-value that corresponds to the upper tail probability of α/2 for each tail. For a two-tailed test, you can use the `scipy.stats.f.ppf` function, which provides the critical value from the F-distribution given the significance level and degrees of freedom.

Here's a Python function that calculates the critical F-value for a two-tailed test:


In [2]:
from scipy.stats import f

def critical_f_value(alpha, df1, df2):
    """
    Computes the critical F-value for a two-tailed test at a given significance level.
    
    Parameters:
    - alpha: Significance level (e.g., 0.05)
    - df1: Degrees of freedom for the numerator
    - df2: Degrees of freedom for the denominator
    
    Returns:
    - critical_value: The critical F-value for the two-tailed test
    """
    # For a two-tailed test, we need to account for the significance level in both tails
    alpha_half = alpha / 2
    
    # Calculate the critical F-value for the upper tail
    critical_value_upper = f.ppf(1 - alpha_half, df1, df2)
    
    return critical_value_upper

# Example usage:
alpha = 0.05
df1 = 10  # Degrees of freedom for the numerator
df2 = 20  # Degrees of freedom for the denominator

critical_value = critical_f_value(alpha, df1, df2)
print(f"Critical F-value: {critical_value}")


Critical F-value: 2.7736713751990805


#### Explanation:
- **Significance Level (alpha)**: This is the threshold for determining the critical value. For a two-tailed test, you divide alpha by 2 to account for both tails of the distribution.
- **Degrees of Freedom (df1, df2)**: These correspond to the numerator and denominator degrees of freedom for the F-distribution.
- **Critical Value Calculation**: Use `f.ppf(1 - alpha_half, df1, df2)` to get the critical value for the upper tail of the F-distribution. For a two-tailed test, the critical value from the upper tail is used, as the F-distribution is not symmetric.

The function returns the critical F-value that you would compare against your calculated F-value to determine statistical significance.


#### Q3. Write a Python program that generates random samples from two normal distributions with known variances and uses an F-test to determine if the variances are equal. The program should output the F-value, degrees of freedom, and p-value for the test.

To write a Python program that generates random samples from two normal distributions with known variances and performs an F-test to determine if the variances are equal, follow these steps:

1. **Generate random samples from two normal distributions.**
2. **Perform an F-test to compare the variances of the two samples.**
3. **Output the F-value, degrees of freedom, and p-value.**

Here's how you can implement this:


In [3]:
import numpy as np
from scipy import stats

def generate_samples(mean1, var1, mean2, var2, n1, n2):
    """
    Generates random samples from two normal distributions.
    
    Parameters:
    - mean1: Mean of the first distribution
    - var1: Variance of the first distribution
    - mean2: Mean of the second distribution
    - var2: Variance of the second distribution
    - n1: Number of samples for the first distribution
    - n2: Number of samples for the second distribution
    
    Returns:
    - samples1: Random samples from the first distribution
    - samples2: Random samples from the second distribution
    """
    samples1 = np.random.normal(mean1, np.sqrt(var1), n1)
    samples2 = np.random.normal(mean2, np.sqrt(var2), n2)
    return samples1, samples2

def f_test_variances(samples1, samples2):
    """
    Performs an F-test to compare the variances of two samples.
    
    Parameters:
    - samples1: Array-like, first sample
    - samples2: Array-like, second sample
    
    Returns:
    - f_value: The F-statistic value
    - p_value: The p-value associated with the F-statistic
    - df1: Degrees of freedom for the numerator
    - df2: Degrees of freedom for the denominator
    """
    # Calculate the variances of the two samples
    var1 = np.var(samples1, ddof=1)
    var2 = np.var(samples2, ddof=1)
    
    # Calculate the F-value
    f_value = var1 / var2 if var1 > var2 else var2 / var1
    
    # Degrees of freedom
    df1 = len(samples1) - 1
    df2 = len(samples2) - 1
    
    # p-value calculation using the F-distribution
    p_value = 1 - stats.f.cdf(f_value, df1, df2)
    
    return f_value, p_value, df1, df2

# Parameters for the distributions
mean1 = 0
var1 = 1
mean2 = 0
var2 = 1
n1 = 50
n2 = 60

# Generate random samples
samples1, samples2 = generate_samples(mean1, var1, mean2, var2, n1, n2)

# Perform the F-test
f_value, p_value, df1, df2 = f_test_variances(samples1, samples2)

# Output the results
print(f"F-value: {f_value:.4f}")
print(f"Degrees of freedom: df1 = {df1}, df2 = {df2}")
print(f"P-value: {p_value:.4f}")


F-value: 1.4359
Degrees of freedom: df1 = 49, df2 = 59
P-value: 0.0918


#### Explanation:

- **Generate Samples**: `generate_samples` generates random samples from two normal distributions using `np.random.normal`. It takes means, variances, and sample sizes as inputs.

- **Perform F-test**: `f_test_variances` calculates the variances of the samples, computes the F-value as the ratio of the variances, and then calculates the p-value using the F-distribution.

- **Output Results**: The program outputs the F-value, degrees of freedom, and p-value.


#### Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from each population. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

##### F-Test for Known Variances

You have two populations with known variances, and you want to test if these variances are significantly different using an F-test. Given:

- Variance of population 1 ($\sigma_1^2$) = 10
- Variance of population 2 ($\sigma_2^2$) = 15
- Sample size for each population ($n_1$ and $n_2$) = 12
- Significance level ($\alpha$) = 0.05

Here's how you can perform the F-test:


In [4]:
import scipy.stats as stats

def f_test_known_variances(var1, var2, n1, n2, alpha):
    """
    Performs an F-test to compare the variances of two populations with known variances.
    
    Parameters:
    - var1: Variance of the first population
    - var2: Variance of the second population
    - n1: Sample size for the first population
    - n2: Sample size for the second population
    - alpha: Significance level
    
    Returns:
    - f_value: The F-statistic value
    - critical_value: Critical F-value from the F-distribution
    - p_value: The p-value associated with the F-statistic
    - df1: Degrees of freedom for the numerator
    - df2: Degrees of freedom for the denominator
    """
    # Calculate the F-value
    f_value = var1 / var2 if var1 > var2 else var2 / var1
    
    # Degrees of freedom
    df1 = n1 - 1
    df2 = n2 - 1
    
    # Critical value from the F-distribution
    critical_value = stats.f.ppf(1 - alpha / 2, df1, df2)
    
    # p-value calculation
    p_value = 2 * (1 - stats.f.cdf(f_value, df1, df2))
    
    return f_value, critical_value, p_value, df1, df2

# Given values
var1 = 10
var2 = 15
n1 = 12
n2 = 12
alpha = 0.05

# Perform the F-test
f_value, critical_value, p_value, df1, df2 = f_test_known_variances(var1, var2, n1, n2, alpha)

# Output the results
print(f"F-value: {f_value:.4f}")
print(f"Critical F-value: {critical_value:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Degrees of freedom: df1 = {df1}, df2 = {df2}")


F-value: 1.5000
Critical F-value: 3.4737
P-value: 0.5124
Degrees of freedom: df1 = 11, df2 = 11


#### Explanation:

For known variances, you calculate the F-value as the ratio of the variances, find the critical F-value for the two-tailed test, and compute the p-value. The significance level is split into two tails for the F-distribution.

#### Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25 products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance level to determine if the claim is justified.

##### F-Test for Sample Variance

In this case, you have:

- **Claimed variance (σ²)** = 0.005
- **Sample variance (s²)** = 0.006
- **Sample size (n)** = 25
- **Significance level (α)** = 0.01

The test is to check if the sample variance significantly deviates from the claimed variance.


In [5]:
def f_test_sample_variance(claimed_var, sample_var, n, alpha):
    """
    Performs an F-test to compare the sample variance to a claimed variance.
    
    Parameters:
    - claimed_var: Claimed variance
    - sample_var: Sample variance
    - n: Sample size
    - alpha: Significance level
    
    Returns:
    - f_value: The F-statistic value
    - critical_value: Critical F-value from the F-distribution
    - p_value: The p-value associated with the F-statistic
    - df: Degrees of freedom for the test
    """
    # Calculate the F-value
    f_value = sample_var / claimed_var
    
    # Degrees of freedom
    df = n - 1
    
    # Critical value from the F-distribution
    critical_value = stats.f.ppf(1 - alpha, df, df)
    
    # p-value calculation
    p_value = 1 - stats.f.cdf(f_value, df, df)
    
    return f_value, critical_value, p_value, df

# Given values
claimed_var = 0.005
sample_var = 0.006
n = 25
alpha = 0.01

# Perform the F-test
f_value, critical_value, p_value, df = f_test_sample_variance(claimed_var, sample_var, n, alpha)

# Output the results
print(f"F-value: {f_value:.4f}")
print(f"Critical F-value: {critical_value:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Degrees of freedom: df = {df}")


F-value: 1.2000
Critical F-value: 2.6591
P-value: 0.3294
Degrees of freedom: df = 24


#### Explanation:

For sample variances compared to a claimed variance, you calculate the F-value as the ratio of the sample variance to the claimed variance, then find the critical value for the F-distribution and compute the p-value. Since it's a one-tailed test, the critical value and p-value are based on one tail.


#### Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an F-distribution and calculates the mean and variance of the distribution. The function should return the mean and variance as a tuple.

Sure! The mean and variance of an F-distribution can be computed given the degrees of freedom for the numerator ($d_1$) and denominator ($d_2$).

Here's a Python function that calculates and returns the mean and variance of the F-distribution:

In [6]:
def f_distribution_mean_variance(d1, d2):
    """
    Calculate the mean and variance of an F-distribution.
    
    Parameters:
    - d1: Degrees of freedom for the numerator.
    - d2: Degrees of freedom for the denominator.
    
    Returns:
    - A tuple containing the mean and variance of the F-distribution.
    """
    if d2 <= 2:
        raise ValueError("Degrees of freedom for the denominator must be greater than 2 to calculate variance.")
    
    mean = d2 / (d2 - 2)
    variance = (2 * (d2 ** 2) * (d1 + d1 - 2)) / (d1 * (d2 - 2) ** 2 * (d2 - 4))
    
    return (mean, variance)

# Example usage:
d1 = 5
d2 = 10
mean, variance = f_distribution_mean_variance(d1, d2)
print(f"Mean: {mean}, Variance: {variance}")


Mean: 1.25, Variance: 0.8333333333333334


Explanation:
- **Mean**: The mean of an F-distribution is given by 

  $
  \frac{d_2}{d_2 - 2}
  $

  provided $ d_2 > 2 $.

- **Variance**: The variance of an F-distribution is given by 

  $
  \frac{2 \cdot d_2^2 \cdot (d_1 + d_1 - 2)}{d_1 \cdot (d_2 - 2)^2 \cdot (d_2 - 4)}
  $

  provided $ d_2 > 4 $.

  If $ d_2 \leq 2 $, the variance is not defined, so the function raises an exception in such cases.


#### Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The sample variance is found to be 25. Another random sample of 15 measurements is taken from another normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test at the 10% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances from two normal populations are significantly different, follow these steps:

#### Steps to Perform the F-Test

1. **State the Hypotheses**:
   - Null Hypothesis ($H_0$): The variances of the two populations are equal ($\sigma_1^2 = \sigma_2^2$).
   - Alternative Hypothesis ($H_1$): The variances of the two populations are not equal ($\sigma_1^2 \neq \sigma_2^2$).

2. **Calculate the F-Statistic**: The F-statistic is calculated as the ratio of the two sample variances. The formula is:
   $
   F = \frac{S_1^2}{S_2^2}
   $
   where $S_1^2$ is the sample variance from the first sample and $S_2^2$ is the sample variance from the second sample.

3. **Determine the Critical Value**: The critical value for the F-distribution depends on the significance level ($\alpha$) and the degrees of freedom of the two samples. Since this is a two-tailed test, you will need to find the critical values for $\alpha/2$ and $1 - \alpha/2$.

   - Degrees of freedom for the numerator ($df_1$) = $n_1 - 1$
   - Degrees of freedom for the denominator ($df_2$) = $n_2 - 1$

4. **Compare the F-Statistic to the Critical Values**:
   - If the F-statistic is greater than the upper critical value or less than the lower critical value, reject the null hypothesis.

##### Calculation

Given:
- Sample variance 1 ($S_1^2$) = 25
- Sample size 1 ($n_1$) = 10
- Sample variance 2 ($S_2^2$) = 20
- Sample size 2 ($n_2$) = 15
- Significance level ($\alpha$) = 0.10 (10%)

1. Compute the F-Statistic:
   $
   F = \frac{S_1^2}{S_2^2} = \frac{25}{20} = 1.25
   $

2. Determine Degrees of Freedom:
   - $df_1 = n_1 - 1 = 10 - 1 = 9$
   - $df_2 = n_2 - 1 = 15 - 1 = 14$

3. Find the Critical Values:
   - For a two-tailed test at $\alpha = 0.10$, each tail has $\alpha/2 = 0.05$. You need to find the critical values for $F_{0.05, 9, 14}$ and $F_{0.95, 14, 9}$.

   Using an F-distribution table or an F-distribution calculator:
   - $F_{0.05, 9, 14} \approx 2.74$
   - $F_{0.95, 14, 9} \approx 0.365$

4. Compare the F-Statistic:
   - If $F$ is greater than 2.74 or less than 0.365, reject the null hypothesis.

##### Conclusion

Since 1.25 is neither greater than 2.74 nor less than 0.365, you do not reject the null hypothesis at the 10% significance level. There is not enough evidence to conclude that the variances are significantly different.


#### Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of waiting times at the two restaurants are significantly different, follow these steps:

##### Steps to Perform the F-Test

1. **State the Hypotheses:**

   - **Null Hypothesis (𝐻₀)**: The variances of waiting times at Restaurant A and Restaurant B are equal (𝜎ₐ² = 𝜎_b²).
   - **Alternative Hypothesis (𝐻₁)**: The variances of waiting times at Restaurant A and Restaurant B are not equal (𝜎ₐ² ≠ 𝜎_b²).

2. **Calculate the Sample Variances:**

   - **For Restaurant A:**

     Calculate the sample mean:

     $
     \bar{X}_A = \frac{24 + 25 + 28 + 23 + 22 + 20 + 27}{7} \approx 24.14
     $

     Calculate the sample variance $ S_A^2 $:

     $
     S_A^2 = \frac{\sum (X_i - \bar{X}_A)^2}{n_A - 1}
     $

     $
     S_A^2 \approx \frac{0.02 + 0.74 + 14.95 + 1.30 + 4.56 + 17.18 + 8.08}{6} \approx 7.81
     $

   - **For Restaurant B:**

     Calculate the sample mean:

     $
     \bar{X}_B = \frac{31 + 33 + 35 + 30 + 32 + 36}{6} \approx 32.83
     $

     Calculate the sample variance $ S_B^2 $:

     $
     S_B^2 = \frac{\sum (X_i - \bar{X}_B)^2}{n_B - 1}
     $

     $
     S_B^2 \approx \frac{3.34 + 0.03 + 4.69 + 7.93 + 0.69 + 10.03}{5} \approx 5.34
     $

3. **Calculate the F-Statistic:**

   $
   F = \frac{S_A^2}{S_B^2} = \frac{7.81}{5.34} \approx 1.46
   $

4. **Determine the Critical Value:**

   Degrees of freedom for Restaurant A (𝑑𝑓ₐ) = $ n_A - 1 = 7 - 1 = 6 $

   Degrees of freedom for Restaurant B (𝑑𝑓_b) = $ n_B - 1 = 6 - 1 = 5 $

   For a two-tailed test at $ \alpha = 0.05 $, you will need the critical value for $ F_{0.025,6,5} $ and $ F_{0.975,5,6} $ (since it's two-tailed).

   Using an F-distribution table or calculator:

   $
   F_{0.025,6,5} \approx 5.41
   $

   $
   F_{0.975,5,6} \approx 0.184
   $

5. **Compare the F-Statistic:**

   If $ F $ is greater than 5.41 or less than 0.184, reject the null hypothesis. Since 1.46 is neither greater than 5.41 nor less than 0.184, you do not reject the null hypothesis at the 5% significance level. There is not enough evidence to conclude that the variances of waiting times at the two restaurants are significantly different.


#### Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83; Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of test scores for the two groups are significantly different, follow these steps:

#### Steps to Perform the F-Test

1. **State the Hypotheses:**
   - **Null Hypothesis (H₀)**: The variances of the two groups are equal (σ_A² = σ_B²).
   - **Alternative Hypothesis (H₁)**: The variances of the two groups are not equal (σ_A² ≠ σ_B²).

2. **Calculate the Sample Variances:**
   - Compute the sample variance for each group.

3. **Calculate the F-Statistic:**
   The F-statistic is calculated as the ratio of the two sample variances. The formula is:

   $
   F = \frac{S_A^2}{S_B^2}
   $

   where $S_A^2$ is the sample variance for Group A and $S_B^2$ is the sample variance for Group B.

4. **Determine the Critical Value:**
   The critical value for the F-distribution depends on the significance level (α) and the degrees of freedom of the two samples.

5. **Compare the F-Statistic to the Critical Value:**
   - If the F-statistic is greater than the critical value, reject the null hypothesis.

#### Calculation

**Given:**

- Group A: 80, 85, 90, 92, 87, 83
- Group B: 75, 78, 82, 79, 81, 84
- Significance level (α) = 0.01 (1%)

1. **Compute the Sample Variances**

   **For Group A:**
   - Calculate the sample mean:

     $
     \bar{X}_A = \frac{80 + 85 + 90 + 92 + 87 + 83}{6} \approx 86.17
     $

   - Calculate the sample variance $S_A^2$:

     $
     S_A^2 \approx 19.75
     $

   **For Group B:**
   - Calculate the sample mean:

     $
     \bar{X}_B = \frac{75 + 78 + 82 + 79 + 81 + 84}{6} \approx 79.50
     $

   - Calculate the sample variance $S_B^2$:

     $
     S_B^2 \approx 10.30
     $

2. **Calculate the F-Statistic**

   $
   F = \frac{S_A^2}{S_B^2} \approx \frac{19.75}{10.30} \approx 1.92
   $

3. **Determine the Critical Value**

   Degrees of freedom for Group A ($df_A$) = $n_A - 1 = 6 - 1 = 5$

   Degrees of freedom for Group B ($df_B$) = $n_B - 1 = 6 - 1 = 5$

   For a two-tailed test at α = 0.01, the critical values for $F_{0.005, 5, 5}$ and $F_{0.995, 5, 5}$ are needed.

   Using an F-distribution table or calculator:
   - $F_{0.01, 5, 5} \approx 6.99$
   - $F_{0.99, 5, 5} \approx 0.143$

4. **Compare the F-Statistic**

   If $F$ is greater than 6.99 or less than 0.143, reject the null hypothesis.

   Since 1.92 is neither greater than 6.99 nor less than 0.143, you do not reject the null hypothesis at the 1% significance level. There is not enough evidence to conclude that the variances of test scores between the two groups are significantly different.
